solutions/usaco-593.mdx

```diff
@@ -7,138 +7,230 @@ author: Óscar Garries, Ryan Chou
+				curr = curr.next(s.dir);
+				time++;
+
+				// Check if FJ's been to this patch of grass before.
+				if (visits.containsKey(curr)) {
+					maxRegrow = Math.min(maxRegrow, time - visits.get(curr));
+				}
+
+				// Update the last time FJ has come upon this patch of grass.
+				visits.put(curr, time);
+			}
+		}
 
-const int INF = 1e9;
+		PrintWriter written = new PrintWriter("mowing.out");
+		// Output -1 if FJ has never gone back to the same patch of grass.
+		written.println(maxRegrow == Integer.MAX_VALUE ? -1 : maxRegrow);
+		written.close();
+	}
+}
 
-int ma[2002][2002];
+//BeginCodeSnip{Point Class}
+class Point {
+	int x, y;
 
-int main(){
-	ios_base::sync_with_stdio(0); cin.tie(0);
-	freopen("mowing.in", "r", stdin);
-	freopen("mowing.out", "w", stdout);
+	public Point(int x, int y) {
+		this.x = x;
+		this.y = y;
+	}
 
-	for (int i = 0; i < 2002; i++) {
-		for (int j = 0; j < 2002; j++) {
-			ma[i][j] = INF;
+	/**
+	 * @return The point reached after moving in the given direction from this point.
+	 */
+	public Point next(char dir) {
+		switch (dir) {
+			case 'N':
+				return new Point(x, y + 1);
+			case 'W':
+				return new Point(x - 1, y);
+			case 'E':
+				return new Point(x + 1, y);
+			case 'S':
+				return new Point(x, y - 1);
+			default:
+				return this;
 		}
 	}
 
-	int sol = INF;
-	bool is = false;
-	int t = 0;
-	int y = 1000, x = 1000;
-	int n;
-	cin >> n;
-	for (int i = 0; i < n; i++) {
-		char direc;
-		int move;
-		cin >> direc >> move;
-		auto upd = [&] (int &cordinate, bool sum) {
-			for (int sum = 1; sum <= move; sum++) {
-				t++;
-				if (sum) cordinate++;
-				else cordinate--;
-				if (ma[y][x] != INF) {
-						is = true;
-						sol = min(sol, t - ma[y][x]);
-					}
-				ma[y][x] = t;
-			}
-		};
-
-		if (direc == 'N') {
-			upd(y, false);
-		} else if (direc == 'E'){
-			upd(x, true);
-		} else if (direc == 'W'){
-			upd(x, false);
-		} else {
-			upd(y, true);
+	@Override
+	public String toString() {
+		return String.format("(%d, %d)", x, y);
 	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(x, y);
+	}
+
+	@Override
+	public boolean equals(Object obj) {
+		return obj instanceof Point && x == ((Point) obj).x && y == ((Point) obj).y;
 	}
-	if (!is) cout << -1 << '\n';
-	else cout << sol << '\n';
 }
+//EndCodeSnip
 ```
 
-</CPPSection>
+</JavaSection>
+<PySection>
+
+```py
+with open('mowing.in') as read:
+	n = int(read.readline())
+	steps = []
+	for _ in range(n):
+		dir_, num_steps = map(str, read.readline().split())
+		steps.append((dir_, int(num_steps)))
+
+# FJ's starting position.
+curr = (0, 0)
+# Make a hashmap, and mark off steps.
+# We'll first mark off his starting position.
+visits = {curr: 0}
+
+time = 0
+max_regrow = float('inf')
+for dir_, num_steps in steps:
+	# Get the direction Farmer John is moving.
+	if dir_ == 'N':
+		delta = 0, 1
+	elif dir_ == 'W':
+		delta = -1, 0
+	elif dir_ == 'E':
+		delta = 1, 0
+	elif dir_ == 'S':
+		delta = 0, -1
```

Comment:
perhaps make delta more specific: `delta = [0, 1]`, ...

Reply:
wdym by this?

Reply:
like write out the list brackets

this is totally optinoal

Reply:
rn `delta` isn't a list, it's a tuple
don't see any particular advantage to making it a list?